fix(db): refactor filtering with subqueries

vovochka404 opened the pull request:
This refactoring allows correctly apply positive and negative filtering to all sub-queries.


Fixes #285

kolaente commented on pkg/models/task_search.go:
```diff
@@ -34,6 +34,61 @@ import (
+	},
+	"assignees": {
+		Table:           "task_assignees",
+		BaseFilter:      "tasks.id = task_id",
```
If this is always the same, would it make sense to add this as fallback?

On the other hand, it might be good to have it explicit.

kolaente reviewed (commented):
Looks quite good already, will do an in-depth review later.

Can you add a few tests for the new cases?

vovochka404 commented:
> Looks quite good already, will do an in-depth review later.
> 
> Can you add a few tests for the new cases?

Fixed test and added one more for filtering with `IncludeNulls`. Andd yep, there was a bug in my logic. I thought that builder object is immutable. Thus, had to redesign work with base filter.

kolaente commented on pkg/models/task_collection_test.go:
```diff
@@ -1272,8 +1314,38 @@ func TestTaskCollection_ReadAll(t *testing.T) {
 			want: []*Task{
 				task1,
 				task2,
+				task3,
```
Shouldn't this also depend on `include nulls`?

vovochka404 replied:
Why? Tasks without labels fully comply with the condition "does not contain the specified label"

kolaente replied:
That does make sense, I guess I was just confused by the change in assertions.

kolaente commented on pkg/models/task_search.go:
```diff
@@ -44,6 +99,20 @@ type dbTaskSearcher struct {
+	var cond = builder.
+		Select("1").
+		From(sf.Table).
+		Where(builder.And(builder.Expr(sf.BaseFilter)))
```
No need for `builder.And` if there's always only one condition.

vovochka404 replied:
Fixed

kolaente commented on pkg/models/task_search.go:
```diff
@@ -102,103 +170,46 @@ func convertFiltersToDBFilterCond(rawFilters []*taskFilter, includeNulls bool) (
-				return nil, err
+		subTableFilterParams, ok := subTableFilters[f.field]
+		if ok {
+			// in fact, it's better to skip all unstrict filters when we're filtering with subquery.
```
Is this a todo? Because the comment seems to be different from the code around it.

vovochka404 replied:
Removed comment. It's not valid.

kolaente commented on pkg/models/task_search.go:
```diff
@@ -102,103 +170,46 @@ func convertFiltersToDBFilterCond(rawFilters []*taskFilter, includeNulls bool) (
-			if err != nil {
-				return nil, err
+		subTableFilterParams, ok := subTableFilters[f.field]
+		if ok {
```
Can you return earlier?

vovochka404 replied:
I don't see how to make it better.
